### PR TITLE
[7.4-only] LPS-121853 Create new clay:management-toolbar tag

### PR DESCRIPTION
**Pull request by markocikos**
Fixes https://issues.liferay.com/browse/LPS-121853

In this PR, we are creating the new management toolbar, based on Clay v3, and applying it to `frontend-taglib-clay-sample-web`

Test plan:
1. Place `frontend-taglib-clay-sample-web` on a page and open 'Management Toolbar' tab.
2. The component should contain all features defined in [Lexicon guidelines](https://liferay.design/lexicon/core-components/toolbars/management-bar/).

Test plan (optional):
1. Test usage in blogs, by removing `-v2` on [this line](https://github.com/liferay/liferay-portal/blob/master/modules/apps/blogs/blogs-web/src/main/resources/META-INF/resources/blogs_admin/view_entries.jsp#L34).

Notes:
- The reason to test in blogs is to test selection controls and integration with search container.
- In blogs, bulk actions do not work. This is because as we cannot use `DefaultEventHandler` / `*DisplayContext.getComponentContext()` integration. I'm not sure if there is an easy solution, the only ones I can think of at the moment would include rewriting a lot of blogs implementation. We can leave blogs implementation for a separate task. This is the reason why I didn't include blogs usage update in this PR.
- Initially I did not plan to add server-side rendering for active state, but there are cases when it results in choppy UX. Without SSR we would need a spinner to indicate update is in progress, but that is not consistent with behavior of rest of the page. I believe we should be consistent about SSR, alternatively deciding on far-reaching level that we want to phase it out, with new UX design.
-  I deprecated `advancedSearch`. I don't think this is used anywhere. [Implementation in Clay 2](https://github.com/liferay/clay/blob/2.x/packages/clay-management-toolbar/src/ClayManagementToolbar.soy#L948-L955) looks like a stub. 
- I deprecated `showFiltersDoneButton`. I don't think this is used anywhere. Looking at [Clay 2 Dropdown](https://github.com/liferay/clay/blob/2.x/packages/clay-dropdown/src/ClayDropdownBase.soy#L253-L287), I'm guessing this is an old pattern, to have a button in a dropdown. I am not sure what is 'Done' button suppose to do, as the implementation is left to devs. Close dropdown maybe? Clay 3 dropdown does not support this kind of special button, but instead of adding it I think it's better to just deprecate the attribute.
- Support for 'show more' button in creation menu can be done as a separate task.

/cc @jbalsas @carloslancha @julien 

after

### CODE REVIEW

**Comment by markocikos** [comment]
ci:stop

**Comment by carloslancha** [comment]
Hey @markocikos I'll try to take a look later. Anyway today I'm going on PTO for 3 weeks and Julien wi'll be going too so I guess this will be waiting here for you return :P

**Comment by markocikos** [comment]
ci:test:sf

**Comment by markocikos** [comment]
ci:test:relevant

**Comment by markocikos** [comment]
@jbalsas @carloslancha @julien The PR is finally ready for review :tada: 

I made the last edit to opening comment, removing wip stuff and adding test plan. Also, I made the last force-push, plan to only append commits from now on, unless there will be a massive refactor.

**Comment by julien** [comment]
@markocikos just started reviewing.

Just a few doubts so far:

This is what I see when placing when deploying `frontend-taglib-clay-sample-web` and placing it on a content page



I also changed [this line](https://github.com/liferay/liferay-portal/blob/master/modules/apps/blogs/blogs-web/src/main/resources/META-INF/resources/blogs_admin/view_entries.jsp#L34) you mentioned and deployed `blogs-web`, and this is what I'm seeing



I'm not really sure why yet, but I'll keep on reviewing the code

**Comment by markocikos** [comment]
@julien It looks like Clay icons are incorrect. Don't know why, maybe some caching issue? Could be some `spritemap` wiring is off, but then I would have the same errors.

**Comment by julien** [comment]
@markocikos yeah it's also what i'm thinking, I'm going to do a full `ant all` of that branch while I keep on looking at the code just to be sure.

**Comment by julien** [comment]
@markocikos happy to tell you that after an `ant all` it's OK visually.

**Comment by julien** [comment]
I haven't seen anything suspicious here so I agree on pushing this forward

**Review by carloslancha — COMMENTED** [review]
Just left a couple of inline comments about formatting and naming. About the other stuff:

>The reason to test in blogs is to test selection controls and integration with search container.

`journal-web` is probably a better place to test it since blogs do not integrates the info panel nor have a creation menu.

>In blogs, bulk actions do not work. This is because as we cannot use DefaultEventHandler / *DisplayContext.getComponentContext() integration. I'm not sure if there is an easy solution, the only ones I can think of at the moment would include rewriting a lot of blogs implementation.

Could you please add more details about this?

>I deprecated advancedSearch. I don't think this is used anywhere. Implementation in Clay 2 looks like a stub.

This is something the teams asked for that were present in previous management toolbar (before clay). Don't know if they're using it, if they forgot about it, if it's not needed anymore...

>I deprecated showFiltersDoneButton. I don't think this is used anywhere. Looking at Clay 2 Dropdown, I'm guessing this is an old pattern, to have a button in a dropdown. I am not sure what is 'Done' button suppose to do, as the implementation is left to devs. Close dropdown maybe? Clay 3 dropdown does not support this kind of special button, but instead of adding it I think it's better to just deprecate the attribute.

That was a button to accept the selected filters in the filtering dropdown and reload the dataset with that info. Probably that was developed before taking the decision of leaving filtering dev to each implementer. Just check that is not being used and it's not needed to be sure before removing it.

>Support for 'show more' button in creation menu can be done as a separate task.

Nice! Just keep in mind to not migrate any usage until we have this feature, some teams asked A LOT for it 😅  

After testing it in `journal-web` I found a couple of issues:

- Results bar is always there

image

- Info Panel button doesn't work on first click. It works on the second and subsequent clicks, but then any other button in the MT triggers the panel toggle too

mtjournal

I don't know if these two bugs are because something needs to be properly migrated or because the new MT itself, but worth checking it.

Thanks @markocikos !

File: modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/management_toolbar/ManagementToolbar.js
Comment (markocikos) [reply modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/management_toolbar/ManagementToolbar.js:94]: Currently formatter sets spacing greater than 1 to 1, making 0 or 1 possible. I personally prefer 1 spacing, like in JSPs, not sure if we have a different convention for JS files.

File: modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/servlet/taglib/ManagementToolbarTag.java
Comment (markocikos) [reply modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/servlet/taglib/ManagementToolbarTag.java:1094]: @carloslancha `ManagementToolbarDisplayContext` is used in a few places directly, for example in [ViewFlatUsersDisplayContextFactory](https://github.com/liferay/liferay-portal/blob/master/modules/apps/users-admin/users-admin-web/src/main/java/com/liferay/users/admin/web/internal/display/context/ViewFlatUsersDisplayContextFactory.java#L98) . I don't think it is safe to rename it. We could deprecate it and make a new class with the same content and different name, but that is probably not a good idea, renaming should not be a reason for deprecation.

We can rename internally `_displayContext` to `_managementToolbarModel`, but then we would need to change `displayContext` tag attribute. This would then mess up the very widespread pattern (700+ uses) of setting `*displayContext`s, like [AccountUsersAdminManagementToolbarDisplayContext](https://github.com/liferay/liferay-portal/blob/master/modules/apps/account/account-admin-web/src/main/java/com/liferay/account/admin/web/internal/display/context/AccountUsersAdminManagementToolbarDisplayContext.java) in [this jsp](https://github.com/liferay/liferay-portal/blob/master/modules/apps/account/account-admin-web/src/main/resources/META-INF/resources/account_users_admin/view.jsp#L37). These contexts inherit model properties but also have additional stuff. But, when they are passed to tag, we classify them through `ManagementToolbarDisplayContext` interface, ignoring all the rest. 

Liferay Java naming™ :joy:

Comment (carloslancha) [reply modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/servlet/taglib/ManagementToolbarTag.java:1094]: Starting to ~hate~ get a little bit tired about naming xD

Not sure what to answer here to be honest. I'm going to invoke @jbalsas to see if he can throw some light on what is expected here. Thx Marko!

Comment (julien) [reply modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/servlet/taglib/ManagementToolbarTag.java:1094]: I'm going to be neutral about this: 

I both understand @carloslancha desire to rename this to `ManagementToolbar` to be consistent with our Card tags but I also understand @markocikos comment about the fact that since it's already being used in lots of places we should keep the same name. Sorry guys.

Comment (jbalsas) [reply modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/servlet/taglib/ManagementToolbarTag.java:1094]: Not having too much context, I would defer any renaming until Brian asks for it. Specially if it's an already used API

**Comment by markocikos** [comment]
@carloslancha I added changes in the temporary REVIEW commit. Before forwarding, I'll merge it to the 'create' commit.

Notes:
- Fixed info button toggle init, by placing it in `useEffect`, instead of `onClick`. Moved control to separate file, for readability. There is still a bug where it toggles on all buttons in toolbar, but I ran out of time for today. It is the same code as [here](https://github.com/liferay/liferay-portal/blob/master/modules/apps/item-selector/item-selector-taglib/src/main/resources/META-INF/resources/item_selector_preview/js/ItemSelectorPreview.es.js#L137-L148), it should work :thinking:
- Added `additionalProps` attribute, to support bulk actions. That's the solution to `DefaultEventHandler` / `*DisplayContext.getComponentContext()` integration I mentioned earlier, I just needed to find the existing pattern. If you test now in `journal-web`, actions like item deletion will not work. That is fine, in update task we will need to do something like [this](https://github.com/markocikos/liferay-portal/commit/6c2112b2b1dd31794b7907f063e61e5ac237b97c). That's a commit on my wip branch.
- Added data attribute handling for action dropdown. Moved `get_data_attributes` to parent dir as it is no longer specific to cards.
- Fixed result bar always showing, by additional check if filter labels list is empty. 
- Added spacing between sibling blocks.

**Comment by carloslancha** [comment]
LGTM! Waiting for the info panel bug to be fixed for final review 😄

File: modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/management_toolbar/ManagementToolbar.js
Comment (wincent) [reply modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/management_toolbar/ManagementToolbar.js:96]: Yep, that's all vanilla (unmodified) Prettier behavior. Newlines are basically collapsed in JSX, so there is no difference between 0 and 1 blank lines in a place like this (in the output any number of newlines will get collapsed to a single space), which means you can pick whatever you like best.

**Comment by carloslancha** [comment]
@markocikos rebase conflicts :(

**Comment by markocikos** [comment]
@carloslancha Rebased and fixed conflicts, caused by multiselect tag changes in proximity. There is still an error in sample-web, but it looks like it's unrelated to this PR:
```
2021-01-18 18:01:36.299 ERROR [http-nio-8080-exec-6][IncludeTag:128] Current URL /web/guest generates exception: /partials/form_elements.jsp(146,0) PWC6142: No tag "multiselect" defined in tag library imported with prefix "clay"
org.apache.jasper.JasperException: /partials/form_elements.jsp(146,0) PWC6142: No tag "multiselect" defined in tag library imported with prefix "clay"
	at org.apache.jasper.compiler.DefaultErrorHandler.jspError(DefaultErrorHandler.java:81)
	at org.apache.jasper.compiler.ErrorDispatcher.dispatch(ErrorDispatcher.java:376)

```

Also, fixed a very fun bug that was causing all buttons in the component to open side panel. This was caused by our `SideNavigation` library requiring a [unique element identifier](https://github.com/liferay/liferay-portal/blob/master/modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/side_navigation.es.js#L96-L129), mutating passed note to `id` of element or `id` of ancestor. First ancestor with `id` was component container. I fixed this by adding an `id` to trigger button.

**Comment by julien** [comment]
@markocikos I just fetched this PR, deployed the `frontend-taglib-clay` and `frontend-taglib-clay-sample-web` modules, created a page with the Clay Sample widget but I don't see the error you mention.

The bug @carloslancha concerning the sidebar is fixed.

**Comment by markocikos** [comment]
@carloslancha I created a new task, [Migrate 'show more' feature in management toolbar creation menu](https://issues.liferay.com/browse/LPS-126011), and started working on it. Meanwhile, we can iterate on this PR. If it gets merged quickly, we can also start on usages that don't have the 'show more' feature.

**Comment by carloslancha** [comment]
Nice @markocikos ! LGTM, we can forward this if nobody points any other issue :)

**Review by carloslancha — APPROVED** [review]

**Comment by markocikos** [comment]
Squashed temporary commits and force-pushed, no changes.

**Comment by markocikos** [comment]
ci:forward

**Comment by jbalsas** [comment]
> Nice @markocikos ! LGTM, we can forward this if nobody points any other issue :)

celebrate

**Pull request closed without merging**
